### ISSUE TITLE
Enable all instrumentations by default

### ISSUE THREAD

**Issue by pellared** [open-telemetry/opentelemetry-dotnet-instrumentation#388]

**Comment by nrcventura** [comment]
I think that this is a good thing for us to aim for to improve the experience for our users. When we do this we may also need to provide a way to "disable" or not load certain instrumentations. For now, to simplify things, I think that this should primarily apply to source-based instrumentation. I think that we need to think further about what that means for byte-code based instrumentation.

**Comment by pellared** [comment]
I just want to note that current behavior is strange and not user-friendly.

- Bytecode Instrumentations are automatically enabled.
- Source Instrumentations have to be enabled using `OTEL_DOTNET_AUTO_ENABLED_INSTRUMENTATIONS`.

I find it is so confusing that we should address it even in beta...

Both bytecode instrumentations and source instrumentations can be disabled via `OTEL_DOTNET_AUTO_DISABLED_INSTRUMENTATIONS`.

**Comment by pellared** [comment]
Let's assume that we would like to enable all source instrumentation by default.
How do we want to handle it if it cannot be applied?
E.g. when a console application tries to add ASP.NET Core source instrumentation.
I see the following possibilities:

1. Fail-fast and require the user to set `OTEL_DOTNET_AUTO_DISABLED_INSTRUMENTATIONS=AspNet`. Or explicitly set only the needed source instrumentation via `OTEL_DOTNET_AUTO_ENABLED_INSTRUMENTATIONS`. This gives the user more control over what is going on, reduces warning errors but would require more time for initial setup.
2. Catch exceptions when source instrumentation cannot be applied and simply log a warning. Faster to adopt but if something goes wrong with something that should work - it may be not noticed - because the more source instrumentation will be added the more warning logs would obfuscate the logs.

Of course, both approaches could be implemented and controlled using e.g. an `OTEL_DOTNET_AUTO_FAILFAST` env var.

Any other ideas?

**Comment by nrcventura** [comment]
I see the biggest blocker to auto-loading all source instrumentations being the dependency management issue. The byte-code instrumentation doesn't have explicit dependencies so if we load those instrumentation libraries it won't cause a problem. If I am remembering correctly, once you load an assembly, and it has some errors, I don't think that there's a good away to unload that assembly. Furthermore, if the user is also using some sort of assembly scanning library (typically to support plugins or some other dynamic loading scenario), the simple act of expecting the types in that assembly can potentially trigger exceptions and crash the app if the required dependencies are not available.

To enable "safer" automatic loading, we may need to consider some assembly version checking similar to what is already being done for byte-code instrumentation. Doing that probably won't eliminate the possibility of errors, but it might mitigate the majority of those errors.

**Comment by RassK** [comment]
Linking https://github.com/open-telemetry/opentelemetry-dotnet/issues/3253

**Comment by RassK** [comment]
References to rough PoC are here > 
https://github.com/open-telemetry/opentelemetry-dotnet/issues/3253#issuecomment-1149733278

**Comment by RassK** [comment]
Multiple possible conflicting zones: 

![activity_timeseries](https://user-images.githubusercontent.com/4929112/173875280-37997544-14a5-4c74-b6d0-3111c7778364.jpg)

**Comment by RassK** [comment]
Based on this proposed idea, I'm changing the way it was originally done.
https://github.com/dotnet/runtime/issues/70727#issuecomment-1157113650

The PoC is updated with a new idea. The new idea contains:
* At the startup all instrumentations are enabled if user doesn't specify otherwise
* Subscribing to the sources and creating instrumentation adapter (?) must be separated. 
* At the startup all enabled sources are configured into ActivityListener filter (`ShouldListenTo`)
* When a new assembly is loaded it creates it's own ActivitySource, due filter is already configured it can determine if this source should be listened to.
* Dependency load detector detects the dependency loading point (eg: ASP.NET Core) and creates instrumentation adapter (?). Then the adapter is added into TraceProvider instrumentations list.

The only important part: The extension method that configures sources must not contain reference to the dependency!
In the other word: ASP.NET Core instrumentation setup extension `AddAspNetCoreSources()` must not have a reference to ASP.NET Core framework. Otherwise it will crash.

**Comment by nrcventura** [comment]
Yes, it does seem like "subscribing to sources" and "creating the instrumentation adapter" being separated will help with a solution for this. "Subscrbing to sources" does not seem like it should require types from the assemblies that should be instrumented, but the instrumentation adapters will most likely require the types from those assemblies. That does mean that we can't leverage the extension methods that the instrumentation libraries provide, but I think that is a reasonable trade off.
